Switch from alevin to alevin-fry (with simpleaf)

fmalmeida opened the pull request:
<!--
# nf-core/scrnaseq pull request

Many thanks for contributing to nf-core/scrnaseq!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/scrnaseq/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/scrnaseq/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/scrnaseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

Work on issue #93

fmalmeida commented:
Hi @grst and @rob-p,

Now that we have `SIMPLEAF_INDEX` and `SIMPLEAF_QUANT` modules working, before going on ahead on implementation, I think is good to have a nice review and brainstorm on how the workflow and parameters should look like.

For example, now that `SIMPLEAF_INDEX` also builds the `txp2gene` maps, how we will handle those? How can we reestructure the workflow thinking to check whether we have it from user, from simpleaf or gffread?

https://github.com/nf-core/scrnaseq/blob/07136b05d5b5e9a60ef9568d495ed18dec8ef31d/subworkflows/local/alevin.nf#L48-L56

Also on the module itself, how's best to use chemistry/protocol? How it was done before with the values coming from the function in `lib/`?

All that, so would be good to have a careful review on that.

😄 🚀

rob-p commented:
Hi @fmalmeida,

  Regarding the txp2gene, I think it’s important to use the one that comes from `simpleaf` especially when the user is making use of a _splici_ reference (the recommended default for using `alevin-fry`).  This is because the txp2gene will look different from the standard “2 column” map, in that each transcript is assigned a gene of origin as well as an annotated splicing status (spliced or unspliced).  When the index is made in `—ref` mode (i.e. when the user is not actually invoking `pyroe` to build a _splici_ transcriptome, but instead just indexing a given transcriptome), then we might expect a user-provided `txp2gene` file to be given.  I expect that the former mode of operation will be much more common, but presumably the latter should be supported?


  Regarding the chemistry — `alevin-fry` (and therefore `simpleaf`) supports quite a few as well as a fairly general “generic geometry” that lets it process pretty much any protocol with known barcode and UMI lengths and offsets.  How are the chemistry parameters currently handled in the pipeline among the different tools?

fmalmeida commented:
Hi @rob-p,

> About `txp2gene`

What you said makes total sense. I will adjust to be in that way, to use the one from `simpleaf` but to expect a given one when using given transcripts (`--ref`).

> About chemistry

Great to hear that is very agnostic. But we may require a little more work on better defining it on the pipeline then. For now, this is how chemistries are handled in the pipeline:

https://github.com/nf-core/scrnaseq/blob/07136b05d5b5e9a60ef9568d495ed18dec8ef31d/lib/WorkflowScrnaseq.groovy#L71-L155

fmalmeida commented:
Hi @rob-p ,

Just updated how `txp2gene` is handled base on your propositions. Please see whether you agree:

1. when indexing a built transcriptome with `--transcript_fasta` that is passed as `--refseq` to `simpleaf`, we expect a built `txp2gene` file:

https://github.com/nf-core/scrnaseq/blob/4d3c5c520321a44775d8e30ac82e2db483284f4a/subworkflows/local/alevin.nf#L34-L37

2. Then, we run the pipeline normally. After `simpleaf_index` we then load the correct `txp2gene`. Choosing between the given one or one that comes from `simpleaf`:

https://github.com/nf-core/scrnaseq/blob/4d3c5c520321a44775d8e30ac82e2db483284f4a/subworkflows/local/alevin.nf#L50-L53

apeltzer commented:
Hm, AlevinQC fails: 

```
  Error in checkAlevinInputFiles(baseDir) : 
    Input directory not compatible with Salmon v0.14 or newer (without external whitelist), the following required file(s) are missing or malformed:
  /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/alevin/raw_cb_frequency.txt
  /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/alevin/featureDump.txt
  /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/aux_info/meta_info.json
  /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/aux_info/alevin_meta_info.json
  /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/cmd_info.json
  /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/alevin/whitelist.txt
```

fmalmeida commented:
Yes, I think this is where I stopped last time. It seems that the directory
from simpleaf is not what alevinqc expects, or I did something wrong either
in the command or when selecting what to pass to alevinqc.

But, I didn’t check it further.

On Tue, 13 Sep 2022 at 10.18, Alexander Peltzer ***@***.***>
wrote:

> Hm, AlevinQC fails:
>
>   Error in checkAlevinInputFiles(baseDir) :
>     Input directory not compatible with Salmon v0.14 or newer (without external whitelist), the following required file(s) are missing or malformed:
>   /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/alevin/raw_cb_frequency.txt
>   /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/alevin/featureDump.txt
>   /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/aux_info/meta_info.json
>   /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/aux_info/alevin_meta_info.json
>   /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/cmd_info.json
>   /home/runner/work/scrnaseq/scrnaseq/work/dc/3d69e08ce60e590be85661861d8830/Sample_X_alevin_results/alevin/whitelist.txt
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nf-core/scrnaseq/pull/139#issuecomment-1245062152>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB26UYVKHFA2TZMYHUS7TD3V6A2GXANCNFSM564W2TAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Msc. Felipe Marques de Almeida - Doutorando em Biologia Molecular - UnB.
*Campus* Universitário Darcy Ribeiro, Brasília - CEP 70910-900
Instituto de Ciências Biológicas - Bloco K - 1º andar - Telefone: (61)
31072926

rob-p commented:
Is the latest alevinqc being used?

fmalmeida commented:
Yes, I do think so. The modules are stick to alevinqc version 1.10

rob-p commented:
Let me ping @csoneson here.  There was recently an update to AlevinQC to allow it to work in the case when alevin-fry was used in `unfiltered` mode (the mode being used by default here). Previously, it only worked when knee filtering was used (as some of the graphs didn't make sense in the other case).

csoneson commented:
The current release version of `alevinQC` is [1.12.1](https://bioconductor.org/packages/alevinQC/), the devel version is 1.13.2. The support for processing `alevin-fry` output was added in v1.11.1, so v1.10 is not going to work, unfortunately. Also note that in order to read `alevin-fry` output, you'll need to use the "fry" functions (e.g., `alevinFryQCReport()` rather than `alevinQCReport()`, see https://csoneson.github.io/alevinQC/reference/index.html).

fmalmeida commented:
So I guess this is the problem. Because we are indeed using the latest version available in bioconda, but it turns out that it is not up-to-date with the package.

apeltzer commented:
Seems to be it - we should check on Bioconda whether we can update the recipe easily, then just do that and resolve the issue hopefully :-)

apeltzer commented:
I made a PR for getting 1.12.x on bioconda. https://github.com/bioconda/bioconda-recipes/pull/37106

apeltzer commented:
PR for AlevinQC just got merged to bioconda thanks to @rob-p - we should be able to add a new AlevinQC version then, so this should be fixed then making all tests here pass 👍🏻

fmalmeida commented:
Hi Guys,
If you try to execute now, you will see that we have a different error:

```
  Error in checkAlevinFryInputFiles(mapDir = mapDir, permitDir = permitDir,  : 
    Input directory not compatible with alevin-fry v0.5.0 or newer, the following required file(s) are missing or malformed:
  /workspace/scrnaseq/test/work/f6/092cf1a7152755450e2f16dd72a421/Sample_X_alevin_results/all_freq.bin
  /workspace/scrnaseq/test/work/f6/092cf1a7152755450e2f16dd72a421/Sample_X_alevin_results/permit_freq.bin
```

Which is related to generating this required permitDir. But, I need some help on how to better implementing this on the workflow as [the docs](https://alevin-fry.readthedocs.io/en/stable/generate_permit_list.html) I've found explain it to V2 chemistries, but I guess we need to be agnostic.

Can you help me on implementing this final adjustments?

😬

Khajidu commented:
Hello,
I am testing the alevin-fry branch and came across this error message at the indexing step:
`Error: pyroe failed to return succesfully ExitStatus(unix_wait_status(256))`
This might help you determine what is going wrong.

Khajidu commented:
Oh I found an inconsistency in the `/modules/locale/simpleaf_quant.nf` file (at the very beginning, lines 5-8):

 ```
conda (params.enable_conda ? 'bioconda::simpleaf=0.4.0' : null)
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/simpleaf:0.5.1--h9f5acd7_0' :
        'quay.io/biocontainers/simpleaf:0.5.1--h9f5acd7_0' }"
```

The version should be the same everywhere, right?

fmalmeida commented:
Hi @Khajidu ,

yes, they should be the same.

Khajidu commented:
Oh, same inconsistency in simpleaf_index.nf...

fmalmeida commented:
Now, to finalise this, we should address this:

https://github.com/nf-core/scrnaseq/pull/139#issuecomment-1258005319

Anyone has an idea?

Khajidu commented:
Tested the pipeline again just in case, and still have the same error:

> Hello, I am testing the alevin-fry branch and came across this error message at the indexing step: `Error: pyroe failed to return succesfully ExitStatus(unix_wait_status(256))` This might help you determine what is going wrong.

fmalmeida commented:
I have no idea on such error.

How are you calling the pipeline? Using docker, conda or singularity profile? Using the test dataset or your own?

could you try, e.g. using the test and docker profiles?

Khajidu commented:
I use the singlularity profile (same as while testing with kallisto).

fmalmeida commented:
Hi @rob-p ,

Pinging you to check if you have any idea on what this error experienced by @Khajidu may be related to?

Khajidu commented:
Test profile worked at indexing step, but failed to submit one of the two counting jobs.

Khajidu commented:
And now that I removed cached files, the samplesheet check also failed.

fmalmeida commented:
Very funny that i did not work. I just tried again and it worked until the error commented here https://github.com/nf-core/scrnaseq/pull/139#issuecomment-1258005319

```bash
gitpod /workspace/scrnaseq/test (alevin-fry) $ nextflow run ../main.nf -profile test,docker --outdir results_simpleaf --aligner alevin
Picked up JAVA_TOOL_OPTIONS:  -Xmx3489m
Picked up JAVA_TOOL_OPTIONS:  -Xmx3489m
N E X T F L O W  ~  version 22.04.0
Launching `../main.nf` [jolly_gilbert] DSL2 - revision: 7b2f2dd260

WARN: Found unexpected parameters:
* --simpleaf_rlen: 91
- Ignore this warning: params.schema_ignore_params = "simpleaf_rlen" 



------------------------------------------------------
                                        ,--./,-.
        ___     __   __   __   ___     /,-._.--~'
  |\ | |__  __ /  ` /  \ |__) |__         }  {
  | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                        `._,._,'
  nf-core/scrnaseq v2.0.1dev
------------------------------------------------------
Core Nextflow options
  runName                   : jolly_gilbert
  containerEngine           : docker
  launchDir                 : /workspace/scrnaseq/test
  workDir                   : /workspace/scrnaseq/test/work
  projectDir                : /workspace/scrnaseq
  userName                  : gitpod
  profile                   : test,docker
  configFiles               : /workspace/scrnaseq/nextflow.config

Input/output options
  input                     : https://github.com/nf-core/test-datasets/raw/scrnaseq/samplesheet-2-0.csv
  outdir                    : results_simpleaf

Reference genome options
  genome_fasta              : https://github.com/nf-core/test-datasets/raw/scrnaseq/reference/GRCm38.p6.genome.chr19.fa
  gtf                       : https://github.com/nf-core/test-datasets/raw/scrnaseq/reference/gencode.vM19.annotation.chr19.gtf

Kallisto/BUS Options
  bustools_correct          : true

Institutional config options
  config_profile_name       : Test profile
  config_profile_description: Minimal test dataset to check pipeline function

Max job request options
  max_cpus                  : 2
  max_memory                : 6.GB
  max_time                  : 6.h

!! Only displaying parameters that differ from the pipeline defaults !!
------------------------------------------------------
If you use nf-core/scrnaseq for your analysis please cite:

* The nf-core framework
  https://doi.org/10.5281/zenodo.4643461

* Software dependencies
  https://github.com/nf-core/scrnaseq/blob/master/CITATIONS.md
------------------------------------------------------
executor >  local (12)
[b1/4cd473] process > NFCORE_SCRNASEQ:SCRNASEQ:INPUT_CHECK:SAMPLESHEET_CHECK (samplesheet-2-0.csv)                [100%] 1 of 1 ✔
[07/db5078] process > NFCORE_SCRNASEQ:SCRNASEQ:FASTQC_CHECK:FASTQC (Sample_X)                                     [100%] 2 of 2 ✔
[11/21b9df] process > NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_INDEX (gencode.vM19.annotation.chr19.gtf) [100%] 1 of 1 ✔
[fc/69b37d] process > NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_QUANT (Sample_X)                          [100%] 2 of 2 ✔
[72/760952] process > NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:ALEVINQC (Sample_X)                                [100%] 2 of 2, failed: 2 ✘
[3f/d69355] process > NFCORE_SCRNASEQ:SCRNASEQ:MTX_CONVERSION:MTX_TO_H5AD (Sample_X)                              [100%] 2 of 2 ✔
[-        ] process > NFCORE_SCRNASEQ:SCRNASEQ:MTX_CONVERSION:CONCAT_H5AD                                         -
[-        ] process > NFCORE_SCRNASEQ:SCRNASEQ:MTX_CONVERSION:MTX_TO_SEURAT (Sample_X)                            -
[-        ] process > NFCORE_SCRNASEQ:SCRNASEQ:CUSTOM_DUMPSOFTWAREVERSIONS                                        -
[-        ] process > NFCORE_SCRNASEQ:SCRNASEQ:MULTIQC                                                            -

Error executing process > 'NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:ALEVINQC (Sample_Y)'

Caused by:
  Process `NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:ALEVINQC (Sample_Y)` terminated with an error exit status (1)

Command executed:

  #!/usr/bin/env Rscript
  require(alevinQC)
  alevinFryQCReport(
      mapDir = "Sample_Y_alevin_results/af_map",
      quantDir = "Sample_Y_alevin_results/af_quant",
      permitDir= "Sample_Y_alevin_results",
      sampleId = "Sample_Y",
      outputFile = "alevin_report_Sample_Y.html",
      outputFormat = "html_document",
      outputDir = "./",
      forceOverwrite = TRUE
  )
  
  yaml::write_yaml(
      list(
          'NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:ALEVINQC'=list(
              'alevinqc' = paste(packageVersion('alevinQC'), collapse='.')
          )
      ),
      "versions.yml"
  )

Command exit status:
  1

Command output:
  (empty)

Command error:
  Loading required package: alevinQC
  Registered S3 method overwritten by 'GGally':
    method from   
    +.gg   ggplot2
  Error in checkAlevinFryInputFiles(mapDir = mapDir, permitDir = permitDir,  : 
    Input directory not compatible with alevin-fry v0.5.0 or newer, the following required file(s) are missing or malformed:
  /workspace/scrnaseq/test/work/93/2e0426d631c6121ad6de37bc27497d/Sample_Y_alevin_results/all_freq.bin
  /workspace/scrnaseq/test/work/93/2e0426d631c6121ad6de37bc27497d/Sample_Y_alevin_results/permit_freq.bin
  
  Input directory not compatible with alevin-fry v0.4.3 or newer, the following required file(s) are missing or malformed:
  /workspace/scrnaseq/test/work/93/2e0426d631c6121ad6de37bc27497d/Sample_Y_alevin_results/all_freq.tsv
  
  Calls: alevinFryQCReport -> .alevinQCReport -> checkAlevinFryInputFiles
  Execution halted

Work dir:
  /workspace/scrnaseq/test/work/21/91c20894dbd6fdc07d759b0e88338c

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line


One more CTRL+C to force exit
Adieu
```

Khajidu commented:
Gave it 500G and same error as before, same step.

Khajidu commented:
Dunno about my env config (I always us the cluster-specific profile I mentioned in my kallisto issues) but an update can never be bad.

> Maybe something related to the `singularity` images as from what I just tested with docker on gitpod is working? If that's the case, is there something we can do? Maybe update the bioconda images? Or figure it out whether is some configuration on your env.

Khajidu commented:
I've just noticed something: in my logs, it looks like the pipeline is still calling simpleaf 0.4.0:

```
Error executing process > 'NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_INDEX (sc_gene_models_ncbi_no_genes_no_contigs_notrnas.gtf)'

Caused by:
  Process `NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_INDEX (sc_gene_models_ncbi_no_genes_no_contigs_notrnas.gtf)` terminated with an error exit status (1)

Command executed:

  # export required var
  export ALEVIN_FRY_HOME=.
  
  # prep simpleaf
  simpleaf set-paths
  
  # run simpleaf index
  simpleaf \
      index \
      --threads 6 \
      --fasta sc_ncbi_genome.fasta \
      --gtf sc_gene_models_ncbi_no_genes_no_contigs_notrnas.gtf \
      --rlen 91 \
      -o salmon
  
  **cat <<-END_VERSIONS > versions.yml
  "NFCORE_SCRNASEQ:SCRNASEQ:SCRNASEQ_ALEVIN:SIMPLEAF_INDEX":
      simpleaf: 0.4.0
  END_VERSIONS**

Command exit status:
  1

Command output:
  found `salmon` in the PATH at /usr/local/bin/salmon
  found `alevin-fry` in the PATH at /usr/local/bin/alevin-fry
  found `pyroe` in the PATH at /usr/local/bin/pyroe

Command error:
  Error: pyroe failed to return succesfully ExitStatus(unix_wait_status(256))

Work dir:
  /shared/ifbstor1/projects/bsbii/sc_single_cell_brain/work/36/f3abcf4578f139df74da1e7863d399

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line

```

fmalmeida commented:
Actually, it may not be the case. I just realised the version is hardcoded and not beeing grabbed with the tools command.

😞

fmalmeida commented:
@nf-core-bot fix linting

grst commented on modules/local/simpleaf_index.nf:
```diff
@@ -0,0 +1,48 @@
+process SIMPLEAF_INDEX {
```
Does it make sense for this to go into `nf-core/modules`?

fmalmeida replied:
Hi Greg,
I do think it is. But maybe something we can do after version is released and also after new modules structure is more stable?

grst replied:
fair enough, this would unnecessarily delay this release. Let's aim for that in the next one.

grst commented on modules/local/simpleaf_index.nf:
```diff
@@ -0,0 +1,48 @@
+    
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        simpleaf: 0.4.0
```
should not be hardcoded

fmalmeida replied:
Yes, I forgot to add the command. Should not be hardcoded, indeed. But, we have a problem, the tool does not have a version command.

grst replied:
I see! In general, we store the version in a variable in that case, see end of this section:
https://nf-co.re/developers/modules#general

Or maybe just output the version of salmon instead? The wrapper seems of minor importance.

fmalmeida replied:
If we go for salmon, salmon is being grabbed in the quantification step already :)

fmalmeida replied:
switched so that both simpleaf modules have simplef version in variable as you showed and also get salmon version

grst commented on modules/local/simpleaf_quant.nf:
```diff
@@ -1,11 +1,11 @@
-process SALMON_ALEVIN {
```
Also this one could go to nf-core/modules?

fmalmeida replied:
The same as the other.

grst commented on modules/local/simpleaf_quant.nf:
```diff
@@ -23,25 +23,33 @@ process SALMON_ALEVIN {
+    simpleaf set-paths
+
+    # run simpleaf quant
+    gzip -dcf $whitelist > whitelist.txt
```
this would break if the input file is called `whitelist.txt` (cannot redirect to a file of the same name)

The most efficent way is to have simpleaf read from a subshell (but may not be supported by simpleaf): 
```
   ... \\
   -u <(gzip -dcf $whitelist) \\
```

Another possibility is to append the extension: 
```
gzip -dcf $whitelist > ${whitelist}.uncompressed.txt
```

fmalmeida replied:
just tested, and the tool does not accept subshell.

I will add the suffix.

grst commented on nextflow.config:
```diff
@@ -22,7 +22,8 @@ params {
 
-    // salmon alevin parameters
+    // salmon alevin parameters (simpleaf)
+    simpleaf_rlen     = 91
```
Please add this to the pipeline JSON schema

fmalmeida replied:
Yep, well spotted. When all is fixed in the context of the way the workflow is, I will run the `nf-core tools` to address all changes required in the schema.

fmalmeida replied:
Added

grst commented on subworkflows/local/alevin.nf:
```diff
@@ -31,57 +31,49 @@ workflow SCRNASEQ_ALEVIN {
-        "Must provide either a GTF file ('--gtf') or kallisto gene map ('--kallisto_gene_map') to align with kallisto bustools!"
+    if (transcript_fasta) {
+        assert txp2gene:
+            "Since a built transcript was provided ('--transcript_fasta'), must also provide a simpleaf gene map ('--txp2gene') to use with simpleaf quant!"
```
You probably meant to remove the check for salmon above rather than the one for Kallisto? 

Btw, can kallisto and salmon use the same parameter for the gene map? i.e. is `--kallisto_gene_map` and `--txp2gene` the same thing?

fmalmeida replied:
Actually, I thought on removing and adapting the kallisto one because the alevin workflow does not actually use kallisto bustools, so the asserting was not actually right.

But, I may be wrong on that as well. How do you think the asserting should look like?

And I am not sure about the second question, if they are indeed the same. But, if they are serving the same purpose, I think we can maybe change it to a single parameter, e.g., `--txp2gene`.

grst replied:
Sorry I missed that this is all within the aleving subworkflow, I thought it was in the global workflow. 

In that case you are completely right and it should never have been there. 

> But, if they are serving the same purpose, I think we can maybe change it to a single parameter

agree, but I have no idea. Asked on slack, but that can be addressed in a separate PR if necessary.

grst reviewed (commented):
Looks great! Added a few comments/questions.

fmalmeida commented:
@nf-core-bot fix linting

fmalmeida commented:
Hi @grst,
I think now I have addressed all your comments. Maybe this can be considered ready if you think the way the modules have been 'connected' to the main workflow makes sense.

I have tried to only add the 'main parameters to the module' and let the user play with `ext.args` with wanted. This, is most relevant for `SIMPLEAF_QUANT` where the user can actually set some mutually exclusive parameters for the tool which will then be 'scanned' by the module before constructing the command line.

grst commented on conf/modules.config:
```diff
@@ -70,11 +70,19 @@ if (params.aligner == "alevin") {
                 path: { "${params.outdir}/${params.aligner}" },
                 mode: params.publish_dir_mode
             ]
+            ext.args = "--rlen ${params.simpleaf_rlen}"
```
```suggestion
            ext.args = { "--rlen ${params.simpleaf_rlen}" }
```

just to be sure... if params get overriden in a user config, they might not affect this setting otherwise.

fmalmeida replied:
Hmmm,

Indeed. Think is better to pass --rlen to the module directly?

grst reviewed (approved):
LGTM! Thanks @fmalmeida!

apeltzer commented:
@nf-core-bot fix linting

apeltzer commented:
Ok, lets merge this fellow to `dev` and follow up with a separate PR to fix anything remaining & prepare for a release
